rfcs/0000-move-ecmascript-plugins-to-core.md

```diff
@@ -0,0 +1,158 @@
+- Champion: [Nicolò Ribaudo](https://github.com/nicolo-ribaudo)
+- Implementors: <!-- the names of everyone who will work on the PR. you can leave this empty if you would like someone else to work on it -->
+
+# Summary
+
+This RFC proposes moving all the existing plugins under the `@babel/` namespace that transform standard ECMAScript syntax into the `@babel/core` package.
+
+From a user perspective this change should be almost invisible, but it gives us better stability and great implementation advantages by drastically reducing the ways in which packages at different versions can be mixed together.
+
+# Motivation
+
+One of the biggest problems that our architecture leads to is that users could mix any version of the "core" packages (`@babel/core`, `@babel/traverse`, `@babel/parser`, etc) with any version of the transform plugins (`@babel/plugin-transform-classes`, `@babel/plugin-proposal-object-rest-spread`, etc).
+
+This is particularly problematic when a plugin version is _ahead_ of the core packages versions.
+
+1. When a proposal moves to Stage 4 and is enabled by default in `@babel/parser`, we still need to defensively enable the syntax plugins until the next major release. For example, the `objectRestSpread` plugin was enabled in `@babel/parser@7.1.0` but `@babel/plugin-proposal-object-rest-spread` still needs to inherit from `@babel/plugin-syntax-object-rest-spread` ([source](https://github.com/babel/babel/blob/c22e72eb24b9dd83d43e693067d9c856acf9977d/packages/babel-plugin-proposal-object-rest-spread/src/index.js#L214)) in case someone is using the plugin with `@babel/parser@7.0.0`.
+2. When we introduce a new helper in `@babel/helpers`, we cannot use it safely in any of the transform plugins because users could still have an old `@babel/helpers` dependency. We always need to first check if the helper is defined in the used `@babel/core`/`@babel/helper` version, and fallback to a completely different transformation if it's not ([source](https://github.com/babel/babel/blob/58d2f41930859b5d46001e3ae365ffc09e5a8463/packages/babel-plugin-transform-for-of/src/index.js#L180-L184))
+3. When we introduce a new utility function in `@babel/traverse`, we cannot safely rely on it in the transform plugins and we must defensively handle old `@babel/traverse` versions, either by working around fixed `@babel/traverse` bugs ([source](https://github.com/babel/babel/blob/c22e72eb24b9dd83d43e693067d9c856acf9977d/packages/babel-plugin-proposal-object-rest-spread/src/index.js#L6-L15)) or by inlining the new utility in the plugin itself or in a new `@babel/helper-*` package.
+
+# Detailed design
+
+## API
+
+Currently plugin can be specified in one of the following ways:
+
+- `@babel/plugin-[feature]` or `@babel/[feature]`
+- `@org/babel-plugin-[feature]` or `@org/[feature]`
+- `babel-plugin-[feature]` or `[feature]`
+- `module:[package name]`
+
+We can introduce a new protocol similar to `module:`:
+
+- `internal:[feature]`
```

Comment:
Maybe it should be `builtin:[feature]`?

Reply:
To minimize the adoption friction, I propose we implement its opposite: `"external:"`, which means always check node modules instead of builtin plugins in `@babel/core`.

So, if users upgrade from `@babel/core` from 8.x to 8.y, they don't have to modify their config to opt in the builtin plugins. And for some users, their Babel config is inside node_modules which make it harder to edit, and they have to wait for the whole ecosystem to migrate to builtin plugins.

This happens whenever a stage 3 feature gets advanced. Therefore I suggest we implement `"external:[feature]"` instead, users can opt out to the builtin features if they insist using the unmaintained plugins.

Reply:
So I have thought for a while about this, but I still prefer `internal:` for a few reasons:
- Completely changing the resolution process for a specific plugin in version can cause confusion: we already have multiple ways of specifying plugins, that are then normalized/resolved, but at least they are deterministic. I would prefer avoiding one that depends on which specific `@babel/core` version.
- When a proposal becomes standard, we want to rename it from `-proposal-` to `-transform-`. We have not done this in Babel 8 to avoid renaming the package, and it has caused confusion.
- I don't think that "being stuck on an older version" is too bad: usually changes to plugins are minimal once they become standard.
- If we want the "automatic migration" to happen, whenever we move a plugin to `@babel/core` we can release a new version of the `-proposal-` plugin that starts like this:
  ```js
  if (api.version >= 7.x.y) return { inherits: "internal:name-of-the-internal-plugin" };
  ```
  and we can thus still easily provide the "automatically use the internal plugin" behavior.